src/bmp_matcher.rs

```diff
@@ -0,0 +1,305 @@
+	/// However, if the length of the error `Vec` is not 0, you should consider the results
+	/// potentially incomplete.
+	///
+	/// The `index` matcher *includes* devices that errored when attempting to match them.
+	pub fn find_matching_probes(&self) -> BmpMatchResults
+	{
+		let mut results = BmpMatchResults {
+			found: Vec::new(),
+			filtered_out: Vec::new(),
+			errors: Vec::new(),
+		};
+
+		let devices = match list_devices() {
+			Ok(d) => d,
+			Err(e) => {
+				results.errors.push(e.into());
+				return results;
+			},
+		};
+
+		// Filter out devices that don't match the Black Magic Probe's vid/pid in the first place.
+		let devices = devices.filter(|dev| {
+			let vid = dev.vendor_id();
+			let pid = dev.product_id();
+			BmpPlatform::from_vid_pid(Vid(vid), Pid(pid)).is_some()
+		});
+
+		devices
+			.enumerate()
+			.map(|(index, device_info)| self.matching_probe(index, device_info))
+			.collect()
+	}
+
+	fn matching_probe(&self, index: usize, device_info: DeviceInfo) -> MatchResult
+	{
+		// Consider the serial to match if it equals that of the device or if one was not specified at all.
+		let serial_matches = self.serial.as_deref().is_none_or(|s| Some(s) == device_info.serial_number());
+
+		// Consider the index to match if it equals that of the device or if one was not specified at all.
+		let index_matches = self.index.is_none_or(|needle| needle == index);
+
+		// Consider the port to match if it equals that of the device or if one was not specified at all.
+		let port_matches = self.port.as_ref().is_none_or(|p| {
+			let port = PortId::new(&device_info);
+
+			p == &port
+		});
+
+		// Finally, check the provided matchers.
+		if index_matches && port_matches && serial_matches {
+			match BmpDevice::from_usb_device(device_info) {
+				Ok(bmpdev) => MatchResult::Found(bmpdev),
+				Err(e) => MatchResult::Error(e),
+			}
+		} else {
+			MatchResult::NoMatch(device_info)
+		}
+	}
+}
+
+#[derive(Debug, Default)]
+pub struct BmpMatchResults
+{
+	pub found: Vec<BmpDevice>,
+	pub filtered_out: Vec<DeviceInfo>,
+	pub errors: Vec<Report>,
+}
+
+impl FromIterator<MatchResult> for BmpMatchResults
+{
+	fn from_iter<I: IntoIterator<Item = MatchResult>>(iter: I) -> Self
+	{
+		let mut results = BmpMatchResults {
+			found: Vec::new(),
+			filtered_out: Vec::new(),
+			errors: Vec::new(),
+		};
+
+		for match_result in iter {
+			match match_result {
+				MatchResult::NoMatch(device_info) => results.filtered_out.push(device_info),
+				MatchResult::Found(bmpdev) => results.found.push(bmpdev),
+				MatchResult::Error(e) => results.errors.push(e),
+			};
+		}
+
+		results
+	}
+}
+
+impl BmpMatchResults
+{
+	/// Pops all found devices, handling printing error and warning cases.
+	pub fn pop_all(&mut self) -> color_eyre::Result<Vec<BmpDevice>>
+	{
+		match self.found.len() {
+			0 => {
+				// Move all out of filtered_out
+				let mut drained: Vec<DeviceInfo> = std::mem::take(&mut self.filtered_out);
```

Comment:
Given this won't be returned at all, why `std::mem::take()`? it would be better to have this be a mutable reference to the self.filtered_out vec arguably. We're unsure of the intent of this line.

Reply:
The intent (and breaking change) is that previously when there was > 1 filtered, it didn't pop all the items (only with 1 it did pop it off). Now it clears all of it out of the `self.filtered_out` as the intent of the function does

`/// Pops all found devices, handling printing error and warning cases.`